### PR TITLE
ARROW-13576: [C++] Replace ExecNode::InputReceived with ::MakeTask

### PR DESCRIPTION
**Pull request by aocsa**
Instead of immediately pushing to outputs, ExecNodes should package their work into tasks and defer execution to allow more sophisticated external scheduling. This will also relieve pipeline breaking ExecNodes of the responsibility to ensure their work is correctly parallelized.

### CODE REVIEW

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (felipeblazing) [inline cpp/src/arrow/compute/exec/filter_node.cc:110]: Is the purpose of the logic here to allow the FilterNode to still be used in the old way while allowing it schedule a task as well?

File: cpp/src/arrow/compute/exec/project_node.cc
Comment (felipeblazing) [inline cpp/src/arrow/compute/exec/project_node.cc:94]: I take it this is debugging code specific to your test and this will be removed?

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (felipeblazing) [inline cpp/src/arrow/compute/exec/filter_node.cc:112]: Whatever we submit to the executor must be more than just a lambda if it is to encode information that the scheduler needs to use. A cost function and context about what query this belongs to and other metadata might want to be part of what the task includes.

Comment (felipeblazing) [reply cpp/src/arrow/compute/exec/filter_node.cc:112]: I am not saying that we need to include those things now but it would be good to see if there is a way to send this information with the task

**Review by felipeblazing — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (felipeblazing) [inline cpp/src/arrow/compute/exec/filter_node.cc:119]: This feels like you would struggle to assure thread affinity as you went from task to task here.

Comment (felipeblazing) [reply cpp/src/arrow/compute/exec/filter_node.cc:119]: The other proposal was to have the task build up and input_received receive a task with the first node in some plan receiving a task that is something akin to, my task is giving you this data.

**Review by felipeblazing — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/filter_node.cc:102]: Well executor is optional so I think will be maintained to run without executor.

File: cpp/src/arrow/compute/exec/project_node.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/project_node.cc]: sure

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (lidavidm) [reply cpp/src/arrow/compute/exec/filter_node.cc:107]: While that's currently true, I think we want to require an executor in the future (even if the executor is "block and do the work on the calling thread"). See https://github.com/apache/arrow/pull/10845#issuecomment-904986082 and ARROW-13741.

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/filter_node.cc:136]: InputFinished (despite its name) can be called before InputReceived (it's less of an "input is done" and more of "this is how many inputs to expect"). So this won't necessarily work; you'll want something like the AtomicCounter pattern used in other exec nodes.

Comment (aocsa) [reply cpp/src/arrow/compute/exec/filter_node.cc]: Thanks  @lidavidm  I am using AtomicCounters and there is not more race condition issues

File: cpp/src/arrow/compute/exec/exec_plan.cc
Comment (westonpace) [inline cpp/src/arrow/compute/exec/exec_plan.cc:292]: Nit: It doesn't matter today but if this is a general method we may want to loop through all outputs.

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (westonpace) [inline cpp/src/arrow/compute/exec/filter_node.cc:118]: Could this be part of SubmitTask so that this if/else loop doesn't have to belong in every node?

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (westonpace) [inline cpp/src/arrow/compute/exec/filter_node.cc:102]: Could this be part of SubmitTask?

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (westonpace) [inline cpp/src/arrow/compute/exec/filter_node.cc:132]: This also feels like code that will have to be repeated in every node.

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (westonpace) [inline cpp/src/arrow/compute/exec/filter_node.cc:150]: This code feels like it will need to be repeated

File: cpp/src/arrow/util/task_scheduler.h
Comment (westonpace) [inline cpp/src/arrow/util/task_scheduler.h:76]: I'm not really sure what the purpose is here.  I thought "a scheduler has many executors" but these seems like "a scheduler is an executor".  I don't think the scheduler should need `MakeEternal` and there shouldn't be `MakeCpuTaskScheduler`.

Maybe there is a `MakeDefaultScheduler` which has only a single executor (the CPU executor).  The scheduler shouldn't worry about `OwnsThisThread` or `GetCapacity`.

**Review by westonpace — COMMENTED** [review]
Looks like a reasonable start.  I'm not sure I fully understand the task scheduler right now and it would be nice if we could avoid duplication between the nodes on some of the task management work.

File: cpp/src/arrow/util/task_scheduler.h
Comment (felipeblazing) [reply cpp/src/arrow/util/task_scheduler.h:76]: I agree that a scheduler should be able to have multiple executors. It also will eventually have more awareness of the entire system than an executor might have.

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/filter_node.cc:102]: Sure

File: cpp/src/arrow/util/task_scheduler.h
Comment (aocsa) [reply cpp/src/arrow/util/task_scheduler.h]: Agree. TaskScheduler should have multiple executors. I think it can be part of follow-up PR after discussing the next steps.

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/filter_node.cc]: done

**Comment by aocsa** [comment]
I update this PR and I think is ready to review.  I  was struggling trying to work fine with `AsyncTaskGroup` but it seems this util has an issue when concurrent thread are trying to know when `all_task` are `done`.  See here: https://gist.github.com/aocsa/c4e66e9061d1dc9f1a0e8253ffad2be3

So now I am using TaskGroup class in async mode with a `stop_token` to cancel tasks when `ExecNode::StopProducing`  is requested. cc @felipeblazing, @westonpace

File: cpp/src/arrow/util/thread_pool.h
Comment (lidavidm) [inline cpp/src/arrow/util/thread_pool.h:345]: Do we still need to make this public?

File: cpp/src/arrow/compute/exec/exec_plan.cc
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/exec_plan.cc:315]: This is racy/this is a TOC-TOU error.

Comment (lidavidm) [reply cpp/src/arrow/compute/exec/exec_plan.cc:315]: So long as this is only called after checking AtomicCounter, is the check here necessary?

File: cpp/src/arrow/compute/exec/exec_plan.h
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/exec_plan.h:232]: I think all three new methods should be protected (also has_executor can be const, no?)

File: cpp/src/arrow/compute/exec/exec_plan.h
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/exec_plan.h:267]: No need to initialize here if we're also initializing in the constructor.

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/filter_node.cc:141]: Maybe this should be the default implementation in the base class now? (Though I guess for this PR that requires more refactoring.)

Comment (lidavidm) [reply cpp/src/arrow/compute/exec/filter_node.cc:141]: (And in general, if we're codifying patterns like `AtomicCounter batch_count_`, I think more of these boilerplate methods could be implemented in the base class instead, once we're OK with the overall design in this PR.)

**Review by lidavidm — COMMENTED** [review]
Some implementation-level comments. Thanks for finding the error in AsyncTaskGroup; we should figure out what's going on there as well.

**Comment by lidavidm** [comment]
Ah, for AsyncTaskGroup, I think what's happening is that it allows adding tasks even after WaitForTaskToFinish, but it's assumed that this will only be done by another task. But for this PR, presumably that's not the case - we're adding tasks externally so that invariant may be violated, and so running_tasks_ may hit 0, the future may be marked finished, and then a concurrent call may add another task. So the semantics of AsyncTaskGroup don't quite fit what we want here.

**Comment by westonpace** [comment]
FYI, in the new dataset writer node PR, I've renamed `WaitForTaskToFinish` to `End` so it's clear that it should only be called once.  I've also added `OnFinished` which returns a future and can be accessed early.  So you can call `OnFinished` as much as you want (e.g. before you add any tasks even) but you only call `End` once.  I think this clears up the confusion.

Check it out here: https://github.com/apache/arrow/pull/11017/files

If you want to use this I can pull out the changes to AsyncTaskGroup and submit them as a small PR on their own today.  Otherways, it will probably be some ways off before 11017 merges.

**Review by westonpace — COMMENTED** [review]
I like the idea of combining similar logic in one place but I don't think this implementation is not going to make sense for all exec nodes.  Rather than write one huge comment I drafted up a bare sketch of an idea here: https://docs.google.com/document/d/1aJWvU-9fxnl8eE6Iw-IebVVs8wxkZ30f1rNpSBOBgmY/edit#heading=h.qv2yr7aeayxs

I didn't spend too much time on it so feel free to reject it out of the box.  However, I think we will want to either create various ExecNode subclasses or create an "exec node runner" concept.  This way we keep the base ExecNode API pure.  This allows nodes that need to have a rather custom implementation to keep their custom implementation and nodes that can share common logic can do so.

File: cpp/src/arrow/compute/exec/exec_plan.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/exec_plan.cc:315]: Based on my test and profiling test, yes this is necessary. The first call to this functions could when `batch_count_` reach the total_batches, but at the same time ExecNode::StopProducing signal could happens. So only once it's neccesary to mark as Finished  `finished_`.

File: cpp/src/arrow/compute/exec/exec_plan.h
Comment (aocsa) [reply cpp/src/arrow/compute/exec/exec_plan.h:232]: For sure

File: cpp/src/arrow/compute/exec/exec_plan.cc
Comment (lidavidm) [reply cpp/src/arrow/compute/exec/exec_plan.cc:315]: Hmm. I think we have a race, then. The point of AtomicCounter is so that we only ever make one call to MarkFinished. I also just noticed that batch_count_.Increment() needs to go inside the task, not outside it.

**Comment by aocsa** [comment]
I refactor this PR based of your feedbacks. 
Main change:
- Refactor using composition to create two implementations `MakeSimpleSyncRunner` and `MakeSimpleParallelRunner` for the use in filer and project node. 
- Fixed some race condition
- Use of AsyncTaskGroup again.
cc @westonpace, @felipeblazing, @lidavidm

File: cpp/src/arrow/util/future.h
Comment (lidavidm) [inline cpp/src/arrow/util/future.h:853]: nit: maybe `AllFinished`?

File: cpp/src/arrow/util/future.h
Comment (lidavidm) [inline cpp/src/arrow/util/future.h:848]: nit: to be clear, this could be "…status of a failing future" (since "first failing future" above implies the first one to fail, not the first one in the list).

File: cpp/src/arrow/compute/exec/sink_node.cc
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/sink_node.cc:178]: Hmm, is it possible for input_counter_.Cancel() to be false?

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/filter_node.cc:66]: This check could perhaps just be folded into a `MakeSimpleRunner`.

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/filter_node.cc:119]: Why not fold most of this logic into the impl itself?

```cpp
if (!impl_->SubmitTask(task)) {
  StopProducing();
}
```

File: cpp/src/arrow/compute/exec/exec_node_runner_impl.h
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/exec_node_runner_impl.h:33]: Looking at the implementations, I wonder if this shouldn't actually just implement ExecNode? The reason being that most of the boilerplate we want to get rid of is still there right now, and additionally, there's more opportunity for implementer mistakes since you have to remember to call these new methods on the impl, and since the impl class hides the implementation, it's hard to tell at a glance if things are correct anymore. A subclass approach may not compose as well, but I'm not sure there's composition to be had in the first place here, and it would actually remove most of the boilerplate.

File: cpp/src/arrow/compute/exec/exec_node_runner_impl.cc
Comment (westonpace) [inline cpp/src/arrow/compute/exec/exec_node_runner_impl.cc:106]: `this->StopProducing()` is going to call `if (input_counter_.Cancel())` anyways so you can skip this second check.

File: cpp/src/arrow/compute/exec/exec_node_runner_impl.cc
Comment (westonpace) [inline cpp/src/arrow/compute/exec/exec_node_runner_impl.cc:114]: `this->finished_.MarkFinished(Status::OK())` is synonymous with `this->finished_.MarkFinished()` I think (I might be wrong on this) so you could simplify this to...

```
if (!this->finished_.is_finished()) {
  this->finished_.MarkFinished(status);
}
```

File: cpp/src/arrow/compute/exec/exec_node_runner_impl.cc
Comment (westonpace) [inline cpp/src/arrow/compute/exec/exec_node_runner_impl.cc:123]: `RequestStop` won't take hold immediately, it simply tells the task group to stop running any additional tasks but it won't/can't interrupt any currently running tasks.  If we mark finished right away here then there is a risk that you discard the exec plan while tasks are still running.  I think, even in error scenarios, we must wait for the task group to finish before marking `finished_` complete.

**Review by westonpace — COMMENTED** [review]
Some thoughts in addition to David's suggestion of a base class.  I see the challenge of the composition approach better now.  Since the runner doesn't have visibility of the inputs/outputs it's difficult for the runner to take care of things like forwarding signals backwards and forwards.

File: cpp/src/arrow/compute/exec/exec_node_runner_impl.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/exec_node_runner_impl.cc]: it is not correct, this is just a check to not run when finished_ is already marked as Finished

File: cpp/src/arrow/compute/exec/sink_node.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/sink_node.cc:178]: Looks like is possible. Some runs in stress mode (`stress -m 16`) enable `DCHECK(cancelled);`. So a checks is needed.

File: cpp/src/arrow/compute/exec/filter_node.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/filter_node.cc]: I addressed these comments in the refactored MapNode Class

File: cpp/src/arrow/util/future.h
Comment (aocsa) [reply cpp/src/arrow/util/future.h:852]: this part of the code is more PR https://github.com/apache/arrow/pull/11017 which I needed to duplicate here in order to solve some race conditions here.

File: cpp/src/arrow/util/future.h
Comment (aocsa) [reply cpp/src/arrow/util/future.h:847]: same here

File: cpp/src/arrow/compute/exec/exec_node_runner_impl.cc
Comment (aocsa) [reply cpp/src/arrow/compute/exec/exec_node_runner_impl.cc]: `MapNode::MarkFinished()` does not mark finished right away. First It check if it is running in async mode or not, if it is async (has an executor) then it adds a callback  to the `task_group`. Maybe the name is not clear for the porpuse of this logic

**Review by lidavidm — COMMENTED** [review]
Thanks for the updates. I left some small comments but this looks good to me.

File: cpp/src/arrow/compute/exec/exec_plan.h
Comment (aocsa) [reply cpp/src/arrow/compute/exec/exec_plan.h:273]: Yes this is right. 
Not OK status cases are handel at the end of method `SubmitTask`.

**Review by lidavidm — APPROVED** [review]
Thanks. LGTM, I'll let Weston/Felipe comment though.

**Review by felipeblazing — APPROVED** [review]

**Review by westonpace — APPROVED** [review]
Some nits but generally an improvement and removes duplicated code.

**Comment by aocsa** [comment]
Thanks @weston, I rebased this PR and addressed latest feedback. Moreover I ran some benchmarks to see the impact of: 1. the possible issue with ExecBatch copies; 2. async mode execution.  

Note: I runned the benchmark with the [following code ](https://github.com/apache/arrow/blob/3136a9babc8ba5fba55d35b88c7e5a967d4c01e8/cpp/src/arrow/dataset/scanner_benchmark.cc) and with this machine configuration:

```
Run on (16 X 3600 MHz CPU s) with 32Gb RAM
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 512 KiB (x8)
  L3 Unified 16384 KiB (x2)
```

1.1 Sync mode with the lambda task function capturing [batches by copy](https://github.com/apache/arrow/blob/85af59892b83fb49af58c2919d98853b9c1779fd/cpp/src/arrow/compute/exec/exec_plan.h#L311)
```
Benchmark                                               Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------------------
MinimalEndToEndBench/100/10/min_time:1.000           3.46 ms         1.71 ms          812 items_per_second=586.126/s
MinimalEndToEndBench/1000/100/min_time:1.000         52.0 ms         38.0 ms           36 items_per_second=26.3494/s
MinimalEndToEndBench/10000/100/min_time:1.000        1102 ms          997 ms            2 items_per_second=1.00278/s
MinimalEndToEndBench/10000/1000/min_time:1.000       4752 ms         4644 ms            1 items_per_second=0.215319/s
```
1.2 Sync mode with the lambda task function capturing batches by move [with std::bind ](https://gist.github.com/aocsa/9da1f32ae1c36c133316e32d84711bc3) 
```
---------------------------------------------------------------------------------------------------------
Benchmark                                               Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------------------
MinimalEndToEndBench/100/10/min_time:1.000           3.02 ms         1.62 ms          885 items_per_second=617.506/s
MinimalEndToEndBench/1000/100/min_time:1.000         51.7 ms         37.9 ms           35 items_per_second=26.4141/s
MinimalEndToEndBench/10000/100/min_time:1.000        1132 ms         1041 ms            1 items_per_second=0.961052/s
MinimalEndToEndBench/10000/1000/min_time:1.000       4795 ms         4680 ms            1 items_per_second=0.213687/s
```
 
As you can see ExecBatch copies were not the culprit of worst performance in some queries because there aren't "that" many ExecBatch instances and the copy is cheap (most of the case just references). 

2.  Execution without/with ThreadPool
```
without threadpool
------------------------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------------
MinimalEndToEndBench/100/10/0/min_time:1.000            2.33 ms         2.33 ms          601 items_per_second=428.441/s
MinimalEndToEndBench/1000/100/0/min_time:1.000          46.8 ms         46.8 ms           30 items_per_second=21.3571/s
MinimalEndToEndBench/10000/100/0/min_time:1.000         1172 ms         1172 ms            1 items_per_second=0.853482/s
MinimalEndToEndBench/10000/1000/0/min_time:1.000        4906 ms         4905 ms            1 items_per_second=0.203876/s
MinimalEndToEndBench/10000/10000/0/min_time:1.000      52141 ms        52129 ms            1 items_per_second=0.0191832/s

with threadpool
MinimalEndToEndBench/100/10/1/min_time:1.000            3.87 ms         1.87 ms          745 items_per_second=533.584/s
MinimalEndToEndBench/1000/100/1/min_time:1.000          54.1 ms         38.3 ms           37 items_per_second=26.09/s
MinimalEndToEndBench/10000/100/1/min_time:1.000         1153 ms         1010 ms            1 items_per_second=0.990056/s
MinimalEndToEndBench/10000/1000/1/min_time:1.000        4771 ms         4624 ms            1 items_per_second=0.216249/s
MinimalEndToEndBench/10000/10000/1/min_time:1.000      49761 ms        49578 ms            1 items_per_second=0.0201702/s
```
As you can see performance when the workload is small is worst in async mode but when the workload is huge the performance tends to be better.
 

Looking forward your thoughts. cc @felipeblazing

**Review by lidavidm — COMMENTED** [review]
Thanks for the updates! I left some comments on the benchmark.

**Pull request closed without merging**
